### PR TITLE
feat: add repl support for cljs test

### PR DESCRIPTION
**Pull request by yqrashawn**
-------

### Summary

This add REPL support to cljs test

- run `make test` once
- run `yarn run shadow-cljs watch test`
- run `node --require ./test-resources/override.js target/test/test.js --repl` to get the running node process with shadow-cljs node repl
- run tests in REPL with cljs.test, for example [`cljs.test/run-tests`](https://cljs.github.io/api/cljs.test/#run-tests)

status: ready

### CODE REVIEW

File: shadow-cljs.edn [hunk 1]
```diff
@@ -70,17 +70,20 @@
...
-           :optimizations :simple
-           :target        :node-test
-           :closure-defines {status-im.utils.config/INFURA_TOKEN #shadow/env "INFURA_TOKEN"}
+          {:output-to        "target/test/test.js"
+           :output-dir       "target/test"
+           :optimizations    :simple
+           :target           :node-test
+           :main             status-im.test-runner/main
+           ;; set :ui-driven to true to let shadow-cljs inject node-repl
+           :ui-driven        true
```
Comment (yqrashawn) [inline shadow-cljs.edn:79]: shadow-cljs's code that injects the repl and devtools
https://github.com/thheller/shadow-cljs/blob/ecb6f9f236b1305fc8cb1092d4cb2dfc2ca12b4a/src/main/shadow/build/targets/node_test.clj#L47

File: src/status_im/test_runner.cljs [hunk 2]
```diff
@@ -0,0 +1,104 @@
+(ns status-im.test-runner
```
Comment (yqrashawn) [inline src/status_im/test_runner.cljs:1]: This file is mainly copy of https://github.com/thheller/shadow-cljs/blob/ecb6f9f236b1305fc8cb1092d4cb2dfc2ca12b4a/src/main/shadow/test/node.cljs#L1 with some minor changes below

File: src/status_im/test_runner.cljs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  {:dev/always true}
+  (:require
+   [cljs.test :as ct]
+   [clojure.string :as str]
+   [shadow.test :as st]
+   [shadow.test.env :as env]))
+
+(defonce repl? (atom false))
```
Comment (yqrashawn) [inline src/status_im/test_runner.cljs:9]: `@repl?` is `true` when running with `--repl` flag, `node --require ./test-resources/override.js target/test/test.js --repl`

File: src/status_im/test_runner.cljs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+(defmethod ct/report [::ct/default :end-run-tests] [m]
+  (when-not @repl?
```
Comment (yqrashawn) [inline src/status_im/test_runner.cljs:12]: Do not exit when tests ended while running with `--repl`

File: src/status_im/test_runner.cljs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  (reduce
+   (fn [opts arg]
+     (cond
+       (= "--help" arg)
+       (assoc opts :help true)
+
+       (= "--list" arg)
+       (assoc opts :list true)
+
+       (= "--repl" arg)
```
Comment (yqrashawn) [inline src/status_im/test_runner.cljs:32]: parse `--repl` flag

File: src/status_im/test_runner.cljs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                     (fn [m]
+                       (tap> [:test m (ct/get-current-env)])
+                       (prn m))))]
+
+    (cond
+      help
+      (do (println "Usage:")
+          (println "  --list (list known test names)")
+          (println "  --test=<ns-to-test>,<fqn-symbol-to-test> (run test for namespace or single var, separated by comma)")
+          (println "  --repl (start node without automatically running tests)"))
```
Comment (yqrashawn) [inline src/status_im/test_runner.cljs:76]: `--repl` doc

File: src/status_im/test_runner.cljs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        (println "Namespace:" ns)
+        (doseq [var (:vars ns-info)
+                :let [m (meta var)]]
+          (println (str "  " (:ns m) "/" (:name m))))
+        (println "---------------------------------"))
+
+      repl
+      (do
+        (reset! repl? true)
+        (js/process.on "SIGINT" #(js/process.exit 0)))
```
Comment (yqrashawn) [inline src/status_im/test_runner.cljs:91]: handle `--repl` flag
`(js/process.on "SIGINT" #(js/process.exit 0))` is used to handle `C-c`

File: src/status_im/utils/test.cljs [hunk 8]
```diff
@@ -57,4 +57,5 @@
                               (str test-dir "/keystore/" key-uid))))
 
             :identicon (fn [pk]
-                         (.identicon native-status pk))}))
+                         (.identicon native-status pk))
+            :startLocalNotifications identity}))
```
Comment (yqrashawn) [inline src/status_im/utils/test.cljs]: Mock status-go `startLocalNotifications` so that the process won't exit on it

**Review by flexsurfer — APPROVED** [review]

**Comment by erikseppanen** [comment]
Thanks for adding this! 

@yqrashawn I'm not sure how to get the repl up and running and connected to it. Is this supposed to start a node repl that I can connect to? If so, i don't see it after performing the first two steps in your instructions. Is there an additional step I should perform, maybe with shadow-cljs or something? I'm using emacs/cider.

**Comment by yqrashawn** [comment]
> Thanks for adding this! 
> 
> 
> 
> @yqrashawn I'm not sure how to get the repl up and running and connected to it. Is this supposed to start a node repl that I can connect to? If so, i don't see it after performing the first two steps in your instructions. Is there an additional step I should perform, maybe with shadow-cljs or something? I'm using emacs/cider.

Oops there's one step missing. You need to `yarn run shadow-clue watch test` before the node one.

**Pull request merged**